cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go

```diff
@@ -0,0 +1,388 @@
+		return nil, CodedError(400, fmt.Sprintf("Failed to get backup '%v': missing namespace", bkp.Spec.BackupName))
+	}
+
+	// volume name is expected
+	if len(strings.TrimSpace(bkp.Spec.VolumeName)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to get backup '%v': missing volume name", bkp.Spec.BackupName))
+	}
+
+	openebsClient, k8sClient, err := loadClientFromServiceAccount()
+	if err != nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create openEBSClient '%v'", err))
+	}
+
+	bkp.Name = bkp.Spec.SnapName + "-" + bkp.Spec.VolumeName
+	b, err := openebsClient.OpenebsV1alpha1().BackupCStors(bkp.Namespace).Get(bkp.Name, v1.GetOptions{})
+	if err != nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed to fetch backup error:%v", err))
+	}
+
+	if b.Status != v1alpha1.BKPCStorStatusDone && b.Status != v1alpha1.BKPCStorStatusFailed {
+		// check if node is running or not
+		bkpNodeDown := checkIfBKPPoolNodeDown(k8sClient, b)
+		// check if cstor-pool-mgmt container is running or not
+		bkpPodDown := checkIfBKPPoolPodDown(k8sClient, b)
+
+		if bkpNodeDown || bkpPodDown {
+			// Backup is stalled, let's find last-backup status
+			laststat := findLastBackupStat(openebsClient, b)
+			// Update Backup status according to last-backup
+			updateBackupStatus(openebsClient, b, laststat)
+
+			// Get updated backup object
+			b, err = openebsClient.OpenebsV1alpha1().BackupCStors(bkp.Namespace).Get(bkp.Name, v1.GetOptions{})
+			if err != nil {
+				return nil, CodedError(400, fmt.Sprintf("Failed to fetch backup error:%v", err))
+			}
+		}
+	}
+
+	out, err := json.Marshal(b)
+	if err == nil {
+		bOps.resp.Write(out)
+		return nil, nil
+	}
+
+	return nil, CodedError(400, fmt.Sprintf("Failed to encode response data"))
+}
+
+// checkIfBKPPoolNodeDown will check if pool node is running or not
+func checkIfBKPPoolNodeDown(k8sclient *kubernetes.Clientset, bkp *v1alpha1.BackupCStor) bool {
+	var nodeDown = true
+
+	pod := findPodFromCStorID(k8sclient, bkp.Labels["cstorpool.openebs.io/uid"])
+	if pod == nil {
+		glog.Errorf("Failed to find pod for backup:%v", bkp.Name)
+		return nodeDown
+	}
+
+	if pod.Spec.NodeName == "" {
+		return nodeDown
+	}
+
+	node, err := k8sclient.CoreV1().Nodes().Get(pod.Spec.NodeName, v1.GetOptions{})
+	if err != nil {
+		glog.Infof("Failed to fetch node info for backup:%v: %v", bkp.Name, err)
+		return nodeDown
+	}
+	for _, nodestat := range node.Status.Conditions {
+		if nodestat.Type == corev1.NodeReady && nodestat.Status != corev1.ConditionTrue {
+			glog.Infof("Node:%v is not in ready state", node.Name)
+			return nodeDown
+		}
+	}
+	return !nodeDown
+}
+
+// checkIfBKPPoolPodDown will check if pool pod is running or not
+func checkIfBKPPoolPodDown(k8sclient *kubernetes.Clientset, bkp *v1alpha1.BackupCStor) bool {
+	var podDown = true
+
+	pod := findPodFromCStorID(k8sclient, bkp.Labels["cstorpool.openebs.io/uid"])
+	if pod == nil {
+		glog.Errorf("Failed to find pod for backup:%v", bkp.Name)
+		return podDown
```

Comment:
we shouldn't be returning podDown as 'true' on unknown statuses of pod or node

Reply:
ok